File: frontend/src/services/quizService.ts
import { gql } from "@apollo/client/core";
import { apolloClient } from "../apollo-client.ts";
import { userService, type User } from "./userService.ts";

export interface Quiz {
  id: string;
  authorId: string;
  title: string;
  description?: string;
  createdAt: string;
  author: User;
  isPublic: boolean;
}

export const quizService = {
  getQuiz: async (id: string) => {
    const { data } = await apolloClient.query({
      query: gql`
        query GetQuiz($id: ID!) {
          quiz(id: $id) {
            id
            authorId
            title
            description
            createdAt
            isPublic
          }
        }
      `,
      variables: { id },
    });
    return data.quiz;
  },

  getQuizzes: async (authorId?: string) => {
    const { data } = await apolloClient.query({
      query: gql`
        query GetQuizzes($authorId: ID) {
          quizzes(authorId: $authorId) {
            id
            authorId
            title
            description
            createdAt
            isPublic
          }
        }
      `,
      variables: { authorId },
    });
    const quizzesWithAuthors = await Promise.all(
      data.quizzes.map(async (quiz: Quiz) => {
        return {
          ...quiz,
          author: await userService.getUser(quiz.authorId),
        };
      })
    );

    return quizzesWithAuthors;
  },

  createQuiz: async (
    authorId: string,
    title: string,
    isPublic: boolean,
    description?: string
  ) => {
    const { data } = await apolloClient.mutate({
      mutation: gql`
        mutation CreateQuiz(
          $authorId: ID!
          $title: String!
          $description: String
          $isPublic: Boolean
        ) {
          createQuiz(
            authorId: $authorId
            title: $title
            description: $description
            isPublic: $isPublic
          ) {
            id
            authorId
            title
            description
            createdAt
            isPublic
          }
        }
      `,
      variables: { authorId, title, description, isPublic },
    });
    return data.createQuiz;
  },

  updateQuiz: async (id: string, title?: string, description?: string) => {
    const { data } = await apolloClient.mutate({
      mutation: gql`
        mutation UpdateQuiz($id: ID!, $title: String, $description: String) {
          updateQuiz(id: $id, title: $title, description: $description) {
            id
            authorId
            title
            description
            createdAt
            isPublic
          }
        }
      `,
      variables: { id, title, description },
    });
    return data.updateQuiz;
  },

  deleteQuiz: async (id: string) => {
    const { data } = await apolloClient.mutate({
      mutation: gql`
        mutation DeleteQuiz($id: ID!) {
          deleteQuiz(id: $id)
        }
      `,
      variables: { id },
    });
    return data.deleteQuiz;
  },
};

File: flutter/lib/data/services/quiz_service.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:graphql_flutter/graphql_flutter.dart';
import 'package:quiz_app/data/services/user_service.dart';
import '../lib/models/quiz_model.dart';
import 'auth_service.dart';

class QuizService {
  final UserService userService = UserService();
  final GraphQLClient _client =
      AuthService().client; // Get client from AuthService
  final secureStorage = const FlutterSecureStorage();

  Future<Context?> _getHeadersContext() async {
    final String? token = await secureStorage.read(
      key: "jwt",
    );

    if (token != null && token.isNotEmpty) {
      return Context.fromList([
        HttpLinkHeaders(
          headers: {
            'Authorization': 'Bearer $token',
          },
        ),
      ]);
    }
    return null;
  }

  Future<List<Quiz>> getQuizzes({String? authorId}) async {
    final String query = """
      query GetQuizzes(\$authorId: ID) {
        quizzes(authorId: \$authorId) {
          id
          authorId
          title
          description
          createdAt
          isPublic
        }
      }
    """;

    final QueryOptions options = QueryOptions(
      document: gql(query),
      variables: {'authorId': authorId},
      fetchPolicy: FetchPolicy.networkOnly, // Or cache policies as needed
      context: await _getHeadersContext(), // Add context with token
    );

    final QueryResult result = await _client.query(options);

    if (result.hasException) {
      print(result.exception.toString());
      throw Exception('Failed to load quizzes');
    }

    final List<dynamic> quizzesJson = result.data?['quizzes'] ?? [];
    for (var quizJson in quizzesJson) {
      quizJson["author"] = await userService.getUser(id: quizJson["authorId"]);
    }
    return quizzesJson.map((json) => Quiz.fromJson(json)).toList();
  }

  Future<Quiz?> getQuiz(String id) async {
    final String query = """
      query GetQuiz(\$id: ID!) {
        quiz(id: \$id) {
          id
          authorId
          title
          description
          createdAt
          isPublic
        }
      }
    """;

    final QueryOptions options = QueryOptions(
      document: gql(query),
      variables: {'id': id},
      fetchPolicy: FetchPolicy.networkOnly,
      context: await _getHeadersContext(), // Add context with token
    );

    final QueryResult result = await _client.query(options);

    if (result.hasException) {
      print(result.exception.toString());
      // Handle specific errors, e.g., not found vs. network error
      throw Exception('Failed to load quiz');
    }

    if (result.data?['quiz'] == null) {
      return null; // Quiz not found
    }

    return Quiz.fromJson(result.data!['quiz']);
  }
}
